tt_metal/impl/program/program.hpp

```diff
@@ -202,7 +211,7 @@ class Program {
     friend detail::Internal_;
 
     const ProgramTransferInfo &get_program_transfer_info() const noexcept;
-    const std::shared_ptr<Buffer> &get_kernels_buffer() const noexcept;
+    const std::shared_ptr<Buffer> get_kernels_buffer(Device* device) const noexcept;
```

Comment:
It seems a bit weird to me that Program can be enqueue-able to any device, yet it internally stores info across devices. It feels like this should exist outside the Program in some way.

Reply:
A program instance stores the kernel binary buffers per device, i.e. the lifetime of these buffers is tied to the program. Even if they existed outside the Program class itself, we would likely need to store some sort of handle or pointer to these buffer objects associated with the program.

That seems like essentially the same behaviour, with an added cost of maintaining extra state. Thoughts?

Reply:
If you are returning something by value please don't use const. https://youtu.be/dGCxMmGvocE

Reply:
Updated